src/java.base/share/classes/jdk/internal/classfile/instruction/ConstantInstruction.java

```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * {@return the type of the constant}
+     */
+    TypeKind typeKind();
+
+    /**
+     * Models an "intrinsic constant" instruction (e.g., {@code
+     * iconst_0}).
+     */
+    sealed interface IntrinsicConstantInstruction extends ConstantInstruction
+            permits AbstractInstruction.UnboundIntrinsicConstantInstruction {
+
+        /**
+         * {@return the type of the constant}
+         */
+        @Override
+        default TypeKind typeKind() {
+            return opcode().primaryTypeKind();
+        }
+    }
+
+    /**
+     * Models an "argument constant" instruction (e.g., {@code
+     * bipush}).
+     */
+    sealed interface ArgumentConstantInstruction extends ConstantInstruction
+            permits AbstractInstruction.BoundArgumentConstantInstruction,
+                    AbstractInstruction.UnboundArgumentConstantInstruction {
+
+        @Override
+        Integer constantValue();
+
+        /**
+         * {@return the type of the constant}
+         */
+        @Override
+        default TypeKind typeKind() {
+            return opcode().primaryTypeKind();
+        }
+    }
+
+    /**
+     * Models a "load constant" instruction (e.g., {@code
+     * ldc}).
+     */
+    sealed interface LoadConstantInstruction extends ConstantInstruction
+            permits AbstractInstruction.BoundLoadConstantInstruction,
+                    AbstractInstruction.UnboundLoadConstantInstruction {
+
+        /**
+         * {@return the constant value}
+         */
+        LoadableConstantEntry constantEntry();
+
+        /**
+         * {@return the type of the constant}
+         */
+        @Override
+        default TypeKind typeKind() {
+            return constantEntry().typeKind();
+        }
+    }
+
+    /**
+     * {@return an intrinsic constant instruction}
+     *
+     * @param op the opcode for the specific type of intrinsic constant instruction,
+     *           which must be of kind {@link Opcode.Kind#CONSTANT}
+     */
+    static IntrinsicConstantInstruction ofIntrinsic(Opcode op) {
+        Util.checkKind(op, Opcode.Kind.CONSTANT);
+        if (op.constantValue() == null)
+            throw new IllegalArgumentException(String.format("Wrong opcode specified; found %s, expected xCONST_val", op));
+        return new AbstractInstruction.UnboundIntrinsicConstantInstruction(op);
+    }
+
+    /**
+     * {@return an argument constant instruction}
+     *
+     * @param op the opcode for the specific type of intrinsic constant instruction,
+     *           which must be of kind {@link Opcode.Kind#CONSTANT}
+     * @param value the constant value
+     */
+    static ArgumentConstantInstruction ofArgument(Opcode op, int value) {
+        Util.checkKind(op, Opcode.Kind.CONSTANT);
+        if (op != Opcode.BIPUSH && op != Opcode.SIPUSH)
+            throw new IllegalArgumentException(String.format("Wrong opcode specified; found %s, expected BIPUSH or SIPUSH", op, op.kind()));
```

Comment:
IDEA shows warning here.
Too many arguments for format string (found: 2, expected: 1)

Seems op.kind() parameter is unused in the format string.

Reply:
I'll fix it, thank you.